bin/autodetect/Main.cc

```diff
@@ -279,19 +279,20 @@ int main(int argc, char** argv) {
 
     // The categorizer knows how to assign categories to records
     ml::api::CFieldDataCategorizer categorizer(jobId, fieldConfig, limits, outputChainer,
-                                               fieldDataCategorizerOutputWriter);
+                                               fieldDataCategorizerOutputWriter,
+                                               persistenceManager.get());
```

Comment:
This line is the actual fix.  All the other changes relate to trying to make clearer that this extra argument was not optional in production.

An automated CI test to prove this works is tricky as it would need to wait for a background persistence to trigger and complete.  It is probably better that a test for the presence of categorizer state documents be added to the long running QA tests.

Reply:
> An automated CI test to prove this works is tricky as it would need to wait for a background persistence to trigger and complete.

This turned out to be less tricky than I thought, because we request a background persist when a datafeed that is doing both lookback and real-time transitions from lookback to real-time.  I have added a test that checks that categorization state is persisted in this situation in elastic/elasticsearch#55243 and pleasingly it fails before this C++ PR is merged but succeeds if run against C++ compiled locally.